liblocation/src/main/java/com/mapbox/android/core/location/LocationCallbackTransport.java

```diff
@@ -0,0 +1,112 @@
+package com.mapbox.android.core.location;
+
+import android.location.Location;
+import android.location.LocationListener;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.Message;
+import android.util.Log;
+
+import java.lang.ref.WeakReference;
+
+import static com.mapbox.android.core.location.Utils.isBetterLocation;
+
+/**
+ * A base wrap class for LocationListener, contains the logic of downsample
+ * location update frequent.
+ */
+public class LocationCallbackTransport implements LocationListener {
+  private static final String TAG = "CallbackTransport";
+  private static final int MESSAGE_UPDATE_LOCATION = 0;
+  private static final int MESSAGE_NOT_CACHE_PERIOD = 1;
+  protected final LocationEngineCallback<LocationEngineResult> callback;
+  private Location currentBestLocation;
+  private long fastestInterval = 0;
+  private Location cacheLocation;
+  private final Handler handler;
+
+  LocationCallbackTransport(LocationEngineCallback<LocationEngineResult> callback) {
+    this.callback = callback;
+    handler = new ThresholdHandler(this);
+  }
+
+  static class ThresholdHandler extends Handler {
+    private final WeakReference<LocationCallbackTransport> weakReference;
+
+    ThresholdHandler(LocationCallbackTransport callbackTransport) {
+      weakReference = new WeakReference<>(callbackTransport);
+    }
+
+    @Override
+    public void handleMessage(Message msg) {
+      LocationCallbackTransport callbackTransport = weakReference.get();
+      if (callbackTransport == null) {
+        return;
+      }
+      switch (msg.what) {
+        case MESSAGE_UPDATE_LOCATION:
+          Location location = callbackTransport.cacheLocation;
+          callbackTransport.cacheLocation = null;
+          callbackTransport.sendLocation(location);
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  @Override
+  public void onLocationChanged(Location location) {
+    if (!handler.hasMessages(MESSAGE_UPDATE_LOCATION)) {
+      //haven't send message in the previous fastInterval period, so send it directly.
+      sendLocation(location);
+    } else {
+      if (!handler.hasMessages(MESSAGE_NOT_CACHE_PERIOD)) {
+        cacheLocation = location;
+      }
+    }
+  }
+
+  private void sendLocation(Location location) {
```

Comment:
we're violating api contract here by calling `sendLocation` on the main thread, consider the case when user provided its background thread `Looper` on which they want this callback to be dispatched on.

Reply:
Good catch, then we need to call this method on the same thread  that `onLocationChanged` is called.